session/src/main/java/io/micronaut/session/http/HttpSessionFilter.java

```diff
@@ -81,19 +83,23 @@ public int getOrder() {
     @Override
     public Publisher<MutableHttpResponse<?>> doFilter(HttpRequest<?> request, ServerFilterChain chain) {
         request.setAttribute(HttpSessionFilter.class.getName(), true);
-        for (HttpSessionIdResolver resolver : resolvers) {
-            List<String> ids = resolver.resolveIds(request);
-            if (CollectionUtils.isNotEmpty(ids)) {
-                String id = ids.get(0);
-                Publisher<Optional<Session>> sessionLookup = Publishers.fromCompletableFuture(() -> sessionStore.findSession(id));
-                Flux<MutableHttpResponse<?>> storeSessionInAttributes = Flux
-                    .from(sessionLookup)
-                    .switchMap(session -> {
-                        session.ifPresent(entries -> request.getAttributes().put(SESSION_ATTRIBUTE, entries));
-                        return chain.proceed(request);
-                    });
-                return encodeSessionId(request, storeSessionInAttributes);
+        try {
+            for (HttpSessionIdResolver resolver : resolvers) {
+                List<String> ids = resolver.resolveIds(request);
+                if (CollectionUtils.isNotEmpty(ids)) {
+                    String id = ids.get(0);
+                    Publisher<Optional<Session>> sessionLookup = Publishers.fromCompletableFuture(() -> sessionStore.findSession(id));
+                    Flux<MutableHttpResponse<?>> storeSessionInAttributes = Flux
+                            .from(sessionLookup)
+                            .switchMap(session -> {
+                                session.ifPresent(entries -> request.getAttributes().put(SESSION_ATTRIBUTE, entries));
+                                return chain.proceed(request);
+                            });
+                    return encodeSessionId(request, storeSessionInAttributes);
+                }
             }
+        } catch (IllegalArgumentException e) {
+            return Flux.error(new HttpStatusException(HttpStatus.BAD_REQUEST, e.getMessage()));
```

Comment:
may as well just throw the exception

Reply:
that is what the user complains about. He thinks we should not return a 500 for a `IllegalArgumentException` and we should return a 400 or maybe a 422. 

I think we should fix this in core with a new `ExceptionHandler` for IllegalArgumentException`. What do you think @dstepanov ?